lib/babe/verify_test.go

```diff
@@ -162,6 +164,94 @@ func TestVerificationManager_VerifyBlock_Ok(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestVerificationManager_VerifyBlock_Secondary(t *testing.T) {
+	babeService := createTestService(t, nil)
+	rt, err := babeService.blockState.GetRuntime(nil)
+	require.NoError(t, err)
+
+	cfg, err := rt.BabeConfiguration()
+	require.NoError(t, err)
+
+	cfg.GenesisAuthorities = types.AuthoritiesToRaw(babeService.epochData.authorities)
+	cfg.C1 = 1
+	cfg.C2 = 1
+	cfg.SecondarySlots = 0
+
+	vm := newTestVerificationManager(t, cfg)
+
+	kp, err := sr25519.GenerateKeypair()
+	require.NoError(t, err)
+
+	dig := createSecondaryVRFPreDigest(t, kp, 0, uint64(0), uint64(0), Randomness{})
+
+	// todo (ed) remove after fix (see test below TestCreateSecondaryVRFPreDigestMarshal
+	fmt.Printf("dig %T\n", dig)
+
+	digEnc, err := scale.Marshal(dig)
+	require.NoError(t, err)
+
+	babePreDigest, err := types.DecodeBabePreDigest(digEnc)
+	// todo (ed) remove after fix (see test below TestCreateSecondaryVRFPreDigestMarshal
+	fmt.Printf("decoded %T\n", babePreDigest)
+
+	// create pre-digest
+	preDigest := &types.PreRuntimeDigest{
+		ConsensusEngineID: types.BabeEngineID,
+		Data:              digEnc,
+	}
+
+	// create new block header
+	number := big.NewInt(1)
+	digest := types.NewDigest()
+	err = digest.Add(*preDigest)
+	require.NoError(t, err)
+
+	// create seal and add to digest
+	seal := &types.SealDigest{
+		ConsensusEngineID: types.BabeEngineID,
+		Data:              []byte{0},
+	}
+	require.NoError(t, err)
+
+	err = digest.Add(*seal)
+	require.NoError(t, err)
+
+	header, err := types.NewHeader(common.Hash{}, common.Hash{}, common.Hash{}, number, digest)
+	require.NoError(t, err)
+
+	block := types.Block{
+		Header: *header,
+		Body:   nil,
+	}
+	err = vm.VerifyBlock(&block.Header)
+	require.EqualError(t, err, "failed to verify pre-runtime digest: could not verify slot claim VRF proof")
+}
+
+func TestCreateSecondaryVRFPreDigestMarshal(t *testing.T) {
+	kp, err := sr25519.GenerateKeypair()
+	require.NoError(t, err)
+
+	dig := createSecondaryVRFPreDigest(t, kp, 0, uint64(0), uint64(0), Randomness{})
+	fmt.Printf("dig %T\n", dig)
+
+	require.IsType(t, types.BabeSecondaryVRFPreDigest{}, *dig)
+
+	// This used to be done with dig.Encode()
+	digEnc, err := scale.Marshal(dig)
+	require.NoError(t, err)
+
+	babeDigest := types.NewBabeDigest()
+	err = scale.Unmarshal(digEnc, &babeDigest)
+	require.NoError(t, err)
+	// Why isn't this type types.BabeSecondaryVRFPreDigest?
+	require.IsType(t, types.BabePrimaryPreDigest{}, babeDigest)
+
+	babePreDigest, err := types.DecodeBabePreDigest(digEnc)
+	require.NoError(t, err)
+	// Why isn't this type types.BabeSecondaryVRFPreDigest?
+	require.IsType(t, types.BabePrimaryPreDigest{}, babePreDigest)
+}
```

Comment:
@jimjbrettj @timwu20 I had old code that was encoding with `dig.Encode()`, but that method is now gone, so I'm using scale marshal, but not getting expected results.  I'm I doing something wrong here?

Reply:
`digEnc` is scale encoded `BabeSecondaryVRFPreDigest`.  For it to unmarshal correctly into the `types.BabeDigest` which is a `scale.VaryingDataType` the first byte of the encoding needs to be the index of `BabeSecondaryVRFPreDigest`.

Reply:
I'm trying to setup conditions to test digest type `types.BabeSecondaryVRFPreDigest` in `verifyPreRuntimeDigest` in `lib\babe\verify.go`.  In `TestVerificationManager_VerifyBlock_Secondary` I create digest with `createSecondaryVRFPreDigest`, then I would call `.Encode` on that result to encode it, add the encoded digest to block header and pass that to `VerifyBlock`.  VerifyBlock calls verifyPreRuntimeDigest and the switch in there would use case `types.BabeSecondaryVRFPreDigest`.  

However, now that there is no longer an `.Encode` function for the digest generated by `createSecondaryVRFPreDigest` I tried encoding with scale.Encode and experienced what I've described above.  

That's what I'm trying to accomplish, @timwu20 yes I'm happy to sync to discuss this.

Reply:
I think like TIm said you need to use the `scale.VaryingDataType` and register the digest types with the byte index, as the BABE digests are represented in Rust with an enum type, so when you encode them they're prefixed with a byte index

Reply:
Yes, I've updated this, it's working now.